src/table/use-column-widths.tsx

```diff
@@ -63,26 +64,77 @@ const WidthsContext = createContext<WidthsContext>({
 interface WidthProviderProps {
   visibleColumns: readonly ColumnWidthDefinition[];
   resizableColumns: boolean | undefined;
+  containerRef: React.RefObject<HTMLElement>;
   children: React.ReactNode;
 }
 
-export function ColumnWidthsProvider({ visibleColumns, resizableColumns, children }: WidthProviderProps) {
-  const visibleColumnsRef = useRef<(PropertyKey | undefined)[] | null>(null);
-  const [columnWidths, setColumnWidths] = useState<Record<PropertyKey, number>>({});
+export function ColumnWidthsProvider({ visibleColumns, resizableColumns, containerRef, children }: WidthProviderProps) {
+  const visibleColumnsRef = useRef<PropertyKey[] | null>(null);
+  const containerWidthRef = useRef(0);
+  const [columnWidths, setColumnWidths] = useState<null | Record<PropertyKey, number>>(null);
 
   const cellsRef = useRef<Record<PropertyKey, HTMLElement>>({});
+  const stickyCellsRef = useRef<Record<PropertyKey, HTMLElement>>({});
   const getCell = (columnId: PropertyKey): null | HTMLElement => cellsRef.current[columnId] ?? null;
-  const setCell = (columnId: PropertyKey, node: null | HTMLElement) => {
+  const setCell = (sticky: boolean, columnId: PropertyKey, node: null | HTMLElement) => {
+    const ref = sticky ? stickyCellsRef : cellsRef;
     if (node) {
-      cellsRef.current[columnId] = node;
+      ref.current[columnId] = node;
     } else {
-      delete cellsRef.current[columnId];
+      delete ref.current[columnId];
     }
   };
+  // Imperatively sets width style for a cell avoiding React state.
+  // This allows setting the style as soon container's size change is observed.
+  const setColumnWidthStyle = useStableCallback((columnId: PropertyKey) => {
+    const cellElement = cellsRef.current[columnId];
+    const stickyCellElement = stickyCellsRef.current[columnId];
+    const column = visibleColumns.find(column => column.id === columnId);
+
+    if (resizableColumns && columnWidths) {
+      const isLastColumn = columnId === visibleColumns[visibleColumns.length - 1]?.id;
+      const totalWidth = visibleColumns.reduce((sum, { id }) => sum + (columnWidths[id] || DEFAULT_COLUMN_WIDTH), 0);
+      if (isLastColumn && containerWidthRef.current > totalWidth) {
+        setElementWidth(cellElement, 'width', 'auto');
+      } else {
+        setElementWidth(cellElement, 'width', columnWidths[columnId]);
+      }
+      setElementWidth(cellElement, 'minWidth', column?.minWidth);
+    }
+    if (!resizableColumns) {
+      setElementWidth(cellElement, 'width', column?.width);
+      setElementWidth(cellElement, 'minWidth', column?.minWidth);
+      setElementWidth(cellElement, 'maxWidth', column?.maxWidth);
+    }
+
+    // Synchronize cell and sticky cells widths.
+    setElementWidth(stickyCellElement, 'width', cellElement?.offsetWidth);
```

Comment:
The sticky column widths were synced in the use-sticky-header before. It worked because first the width was set during thead rendering and then useLayoutEffect of the use-sticky-header fired. Now, the widths are set in the useEffect which happens slightly after than it was before. Besides, the widths can be only synced properly after the browser can figure out the widths of the columns that have no explicit value set, that's why the timeout is added.